Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -80,6 +153,87 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::processDescendantLayersFlatteningRequirements()
+{
+    // Traverse all children in depth first, post-order
+    for (auto* child : m_children) {
+        child->processDescendantLayersFlatteningRequirements();
+
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            child->processFlatteningRequirements();
+    }
+}
+
+void TextureMapperLayer::processFlatteningRequirements()
+{
+    if (m_flattenedLayer)
+        m_flattenedLayer = nullptr;
+
+    // Reset transformations as this layer is root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    m_layerTransforms.combined = { };
+    m_layerTransforms.combinedForChildren = { };
+#if USE(COORDINATED_GRAPHICS)
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.futureLocalTransform = { };
+    m_layerTransforms.futureCombined = { };
+    m_layerTransforms.futureCombinedForChildren = { };
+#endif
+    ComputeTransformData data;
+    if (m_state.maskLayer)
+        m_state.maskLayer->computeTransformsRecursive(data);
+    if (m_state.replicaLayer)
+        m_state.replicaLayer->computeTransformsRecursive(data);
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
+
+    Region layerRegion;
+    computeFlattenedRegion(layerRegion, true);
+
```

Comment:
I think we can remove `TextureMapperFlattenedLayer::valid()` if we return early here.
~~~c++
if (layerRegion.isEmpty())
  return;
~~~

Reply:
I misunderstood. Never mind.

Reply:
`valid()` was a bit bad choice for function name. I renamed it to `needsUpdate().` Also added your suggestion for `if (layerRegion.isEmpty())`